### DISCUSSION TITLE
What are the differences between vars.to.regress and latent.vars? Should I use both for a variable?

### DISCUSSION THREAD

**Discussion by denvercal1234GitHub** [satijalab/seurat#4259] [Q&A] [answered]
Hi there, 

1. What is the difference between specifying vars.to.regress in ScaleData() or SCTransform() versus latent.vars in FindMarkers() in Differential Gene Analysis? Is it that vars.to.regress will reduce the impact of the specified variable to the clusterings whereas latent.vars will reduce the impact of specified variable to the cluster gene markers but not the clusters themselves?

2. If I want to reduce the effect of a certain set of genes, should I specify the percent of counts from these genes in both steps?

Note that in my workflow, I already use SCTransform() to regress out percent.mt and ScaleData() to regress out cell cycle gene difference. 

Thank you very much!

**Comment by timoast** [comment, marked as answer]
`vars.to.regress` in `ScaleData()` and `SCTransform()` specifies variables to regress out of the scaled data or sctransform residuals. These values are typically used for PCA dimension reduction, and so would prevent those variables that are regressed out from contributing much to the PCA. 

`latent.vars` in `FindMarkers()` specifies latent variables to include in the differential expression model. The normalized data, rather than scaled data or sctransform residuals is typically used for differential expression.

> If I want to reduce the effect of a certain set of genes, should I specify the percent of counts from these genes in both steps? 

Reduce the effect on what exactly? If you have a set of genes, eg. the cell cycle genes, that you don't want to drive the clustering results, you can regress out those genes (first adding as a module score) before running PCA, or remove those genes from the set of variable features used in the PCA.

> Note that in my workflow, I already use SCTransform() to regress out percent.mt and ScaleData() to regress out cell cycle gene difference.

You shouldn't be combining SCTransform and ScaleData. Please see the documentation for SCTransform as well as the [paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1)

**Reply by denvercal1234GitHub** [reply]
Hi @timoast - Thank you for your explanation and response! 

But I am confused why I cannot combine SCTransform() and ScaleData for regressing out different variables. From this post here, @jaisonj708 suggests we can use SCTransform() to regress out percent.mt variable. Then, using ScaleData() to regress out cell cycle gene scores. 

So in my workflow, I do SCTransform(vars.to.regress = "percent.mt", do.scale=TRUE, do.center=TRUE). Then, I do ScaleData(vars.to.regress = "CC.Difference", features = all.genes, do.scale = FALSE, do.center = FALSE). This is the APRROACH 2 in #3945. 

Would you mind helping me with this confusion?

Thank you again very much for your help!

**Reply by timoast** [reply]
> But I am confused why I cannot combine SCTransform() and ScaleData for regressing out different variables. From this post here

I don't see any link to a post, but I think some of the confusion here might arise from the question of how to get the cell cycle score needed if you want to regress out cell cycle in sctransform.

Basically, you need a two-step procedure where you first compute the cell cycle scores, then use those cell cycle scores in SCTransform. It doesn't really matter how you initially compute those cell cycle scores: you can run NormalizeData / ScaleData or you can run SCTransform. You will then have a cell cycle score for each cell, and you go _back_ (change assay back to RNA if you had used SCTransform to compute the cell cycle score) and run SCTransform with this value as a `vars.to.regress`.

You do _not_ run ScaleData on the SCTransform results as this would overwrite the sctransform residuals that are stored. You would end up with scaled and centered log-transformed corrected counts, rather than sctransform model residuals.

> So in my workflow, I do SCTransform(vars.to.regress = "percent.mt", do.scale=TRUE, do.center=TRUE). Then, I do ScaleData(vars.to.regress = "CC.Difference", features = all.genes, do.scale = FALSE, do.center = FALSE)

You should not do this

**Reply by denvercal1234GitHub** [reply]
Hi @timoast - The post is #3665 

I see. Would you mind informing me whether the workflow below makes sense now? Thank you so much for pointing out the error. 

1. Create all Seurat objects
2. QC and filter each object
3. **Calculate percent.mt on raw counts**
4. **SCTransform(do.scale=FALSE, do.center=FALSE) each Seurat object**
5. **CellCycleScoring() (while active assay is still SCT Assay)**
6. Find cc score differences
7. Change defaultAssay <- 'RNA'
7. **SCTransform(vars.to.regress = c("percent.mt", "CC.Difference"), do.scale=TRUE, do.center=TRUE) each Seurat object** 
8. Merge all Seurat objects (if I want to merge them for pooled analysis)
9. RunPCA()
10. RunUMAP() 
11. FindNeighbors()
12. FindCluster()
13. Switch to RNA Assay
14. Perform DE analysis

**Reply by timoast** [reply]
Yes, that sounds reasonable

**Reply by denvercal1234GitHub** [reply]
Thank you so much! ☺️

**Reply by denvercal1234GitHub** [reply]
Hi @timoast, 

When I followed the workflow above to Step 9 (merging the 2 Seurat objects). The SCT Assay of the merged object has 0 variable features. Would you mind helping me understand whether the revised workflow below makes sense? Thank you very much for your help!

3. Calculate percent.mt on raw counts on each object
4. Merge all Seurat objects **(merge.data = FALSE)**
5. SCTransform(do.scale=FALSE, do.center=FALSE) on merged object
6. CellCycleScoring() (while active assay is still SCT Assay) on merged object
7. Find cc score differences on merged object
8. Change defaultAssay <- 'RNA'
9. SCTransform(vars.to.regress = c("percent.mt", "CC.Difference"), do.scale=TRUE, do.center=TRUE) **on the merged Seurat object**
10. RunPCA()
